game-app/game-core/src/main/java/games/strategy/triplea/ui/TabbedProductionPanel.java

```diff
@@ -46,95 +53,66 @@ static IntegerMap<ProductionRule> getProduction(
-      if (!ruleList.getSecond().isEmpty()) {
-        tabs.addTab(ruleList.getFirst(), new JScrollPane(getRulesPanel(ruleList.getSecond())));
-      }
+      JScrollPane scrollPane = new JScrollPane();
+      scrollPane.setHorizontalScrollBarPolicy(ScrollPaneConstants.HORIZONTAL_SCROLLBAR_NEVER);
+      scrollPane.setVerticalScrollBarPolicy(ScrollPaneConstants.VERTICAL_SCROLLBAR_ALWAYS);
+      tabs.addTab(ruleList.getFirst(), scrollPane);
     }
+    updateTabContent(tabs, ruleLists);
+    tabs.addChangeListener(e -> updateTabContent(tabs, ruleLists));
+
     final JPanel totals = new JPanel();
     totals.add(left);
     totals.add(remainingResources);
-    add(
-        totals,
-        new GridBagConstraints(
-            0,
-            2,
-            1,
-            1,
-            1,
-            1,
-            GridBagConstraints.WEST,
-            GridBagConstraints.NONE,
-            new Insets(8, 8, 0, 12),
-            0,
-            0));
-    done = new JButton(doneAction);
-    add(
-        done,
-        new GridBagConstraints(
-            0,
-            3,
-            2,
-            1,
-            1,
-            1,
-            GridBagConstraints.CENTER,
-            GridBagConstraints.NONE,
-            new Insets(0, 0, 8, 0),
-            0,
-            0));
-    tabs.validate();
-    this.validate();
+
+    final JPanel bottom = new JPanel();
+    bottom.setBorder(BorderFactory.createEmptyBorder(0, 12, 0, 12));
+    bottom.setLayout(new BorderLayout());
+    bottom.add(totals, BorderLayout.WEST);
+    bottom.add(new JLabel("Attack | Defense | Movement"), BorderLayout.EAST);
+    JPanel donePanel = new JPanel();
+    donePanel.setLayout(new BoxLayout(donePanel, BoxLayout.Y_AXIS));
+    donePanel.add(Box.createVerticalGlue());
+    donePanel.add(done);
+    done.setAlignmentX(Component.CENTER_ALIGNMENT);
+    donePanel.add(Box.createVerticalGlue());
+    bottom.add(donePanel, BorderLayout.CENTER);
+    add(bottom, BorderLayout.SOUTH);
   }
 
   private void calculateRowsAndColumns(
       final ProductionTabsProperties properties, final int largestList) {
-    if (properties == null
-        || properties.getRows() == 0
+    if (properties.getRows() == 0
         || properties.getColumns() == 0
         || properties.getRows() * properties.getColumns() < largestList) {
-      final int maxColumns;
+      int maxColumns;
       if (largestList <= 36) {
         maxColumns = Math.max(8, Math.min(12, divideRoundUp(largestList, 3)));
       } else if (largestList <= 64) {
         maxColumns = Math.max(8, Math.min(16, divideRoundUp(largestList, 4)));
       } else {
         maxColumns = Math.max(8, Math.min(16, divideRoundUp(largestList, 5)));
       }
+      // Limit number of columns to the available width. Reserve 64px for borders & scroll bar.
+      final int availableWidth = Util.getScreenSize(dialog).width - 64;
+      maxColumns = Math.min(maxColumns, availableWidth / cellSize.width);
```

Comment:
Minor, but I wonder why not use a new variable at L115 rather than mutating an existing value?

Nitpicky - this could be a place where we can favor immutabilty: [avoid mutability convention](https://github.com/triplea-game/triplea/blob/master/docs/development/reference/code-conventions/java-code-conventions.md#avoid-mutability)

Overall more descriptive variable naming might help how the code flows for the reader. Suggestions:

L105 `maxColumns` -> `desiredColumns`
L114 `availableWidth` -> `availableScreenWidth`

The above suggestions/convention nit are certainly pretty trivial concerns compared to the complexity of the existing code that makes a variable renaming exercise frustrating. Specifically, it's really hard to know what the first `maxColumns` really represents, what those calculations are, and so a better name is hard to determine. A better name would describe perhaps something more regarding the output of those calculations. That would be an improvement to the biggest problem here where it is not clear what that `if/else` block is doing..

Reply:
Done.